tools/hex2dfu.c

```diff
@@ -0,0 +1,395 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Copyright (C) 2015 Simon Newton.
+ */
+
+#include <ctype.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <getopt.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sysexits.h>
+
+#include "constants.h"
+#include "dfu.h"
+#include "utils.h"
+
+typedef struct {
+  char *input_file;
+  uint32_t lower_address;
+  uint32_t upper_address;
+  uint16_t vendor_id;
+  uint16_t product_id;
+  bool help;
+  bool force;
+} Options;
+
+typedef enum {
+  DATA = 0,
+  END_OF_FILE = 1,
+  EXTENDED_SEGMENT_ADDRESS = 2,
+  START_SEGMENT_ADDRESS = 3,
+  EXTENDED_LINEAR_ADDRESS = 4,
+  START_LINEAR_ADDRESS = 5
+} RecordType;
+
+typedef struct {
+  uint8_t byte_count;
+  uint16_t address;
+  uint8_t record_type;
+  uint8_t *data;
+} HexRecord;
+
+static const uint32_t DEFAULT_LOWER_ADDRESS = 0x1d007000;
+static const uint32_t DEFAULT_UPPER_ADDRESS = 0x1d07ffff;
+static const char HEX_SUFFIX[] = ".hex";
+static const char DFU_SUFFIX[] = ".dfu";
+
+uint8_t *g_data = NULL;
+unsigned int g_data_size = 0;
+
+/**
+ * @brief Convert a pair of characters to a byte.
+ */
+bool HexToUInt8(const char *str, uint8_t *output) {
+  *output = 0;
+  if (!isxdigit(str[0]) || !isxdigit(str[1])) {
+    return false;
+  }
+
+  *output = (digittoint(str[0]) * 16) + digittoint(str[1]);
+  return true;
+}
+
+/**
+ * @brief Convert 4 characters to a uint16_t.
+ */
+bool HexToUInt16(const char *str, uint16_t *output) {
+  uint8_t upper, lower;
+  if (!HexToUInt8(str, &upper) || !HexToUInt8(str + 2, &lower)) {
+    return false;
+  }
+  *output = (upper << 8) + lower;
```

Comment:
Okay, I thought we had similar in other bits of the Ja Rule code that we could pull in.